### DISCUSSION TITLE
S3 Gateway use case

### DISCUSSION THREAD

**Discussion by rkunnamp** [versity/versitygw#686] [Ideas]
I just thought of sharing a use case that I have.

I am trying to store data lake files  in S3  (in delta lake table format) and issue queries against it for a user facing analytics app .  I think this is a very common use case these days. The flexibility associated with storing data in S3 is immense.  But the downside of  such a solution is the latency associated with accessing the data from S3 . 

So I was looking for a solution that would help me  implement a read through cache  , that could possibly reduce the number of round trips to S3 , and reduce the latency associated with my queries.

To my surprise there are not many tools out there that help does that. The only one which I could find was aistore from nvidia (https://github.com/NVIDIA/aistore). 

With Versity, since I have posix file system support and S3 gateway support, I am hoping that I could build such a read through cache on the application side on top of Versity. I also want to record  this use case here, so that when Versity grows to an industry standard solution down the line, this is perhaps something that could be considered to be natively supported.

**Comment by benmcclelland** [comment]
> so that when Versity grows to an industry standard solution down the line

I like the confidence!

Let us know if there are any features that would help to enable this use case. This sounds like a great use of the gateway!

**Reply by scaleoutsean** [reply]
I chuckled as I've learned a few lessons and always start with a low level of confidence. 
I played with this today. Added a LRU (RAM + overflow to disk) and it worked well (observed 10% speed-up vs. LAN-based S3, using "toy" tests against VGW on localhost). But I know I don't know, so I am not confident this hack works correctly or is stable in any scenario. For now I'm content to know it can be done and I'll keep an eye on (simple) use cases where using this could be helpful.

**Reply by benmcclelland** [reply]
I think LRU gets a bit tricky in a multi-gateway deployment unless you have a limited cache assigned per gateway, which probably isnt what users would expect for this. I'm interested if your prototype handles this case better than I'm thinking? I think for on disk cache, CLOCK (https://dl.acm.org/doi/10.1145/3593856.3595887) might be interesting to investigate. Not sure how best to coordinate among gateways though.

**Reply by scaleoutsean** [reply]
You're absolutely right, that's why I emphasized "simple use cases" (where LRU could be helpful) and mentioned how one shouldn't have much confidence approaching these difficult (caching) problems.

My use case is to cache repeated reads to objects where S3 back-end is performance-limited and there are normally no updates of any kind to existing objects (e.g. ISOs for multi-VM builds/deployments based on same ISO). 
For a variety of other, complex or difficult cases, it'd take months to get right (and that's way beyond my ability).

From TFP:

> large-scale simulation study using 5307 block and web cache workloads collected in the past two decades... We find that qick (sic!) demotion --- evicting most new objects very quickly --- is critical for cache efficiency.  

That wouldn't work for me, for example, and I doubt it would work better than LRU even for popular caching workloads like data staged for deep learning. For example, AI workloads that read millions of images - you probably want the entire working set in cache, so there's no eviction at all, then let it all get pushed out when you run the next job and all GPU nodes read a new set of images. You'll never need an image from 22 minutes ago and you'll never evict anything from current working set. 

Maybe CLOCK is better "overall", but even then it depends what markets/use cases are supposed to be addressed. As you know in HPC, for example, people often build for 1 use case and don't care that 99 others don't work at all. But in some generic "enterprise" environment, people want to transparently handle 10 apps and they want to write (not just read/cache) as well.... It's hard to tell what's "better" if "winning criteria" aren't specified. 

I don't know what VGW users think, but personally I'd like a simple LRU plugin for static back-end data that doesn't change. That's easy to maintain/support and works well for analytics and repeated reads that can be satisfied from one VGW instance. If you need multiple VGWs for bandwidth/scale-out, every node caches the same data so that's not ideal, but collaborative caching would be very complicated and probably need to consider specific S3 client/app or set of apps, which should be done in a specialized plugin in any case... The 80/20 rule applies here, too, IMO.

**Reply by benmcclelland** [reply]
I guess there might also be some consistency concerns with dedicated caches if one gateway caches an object and later another gateway handles the delete object request.

I like the original idea in this thread of stacking backends for cache purposes. That way the local posix could be shared across any number of gateways the same as it does now. The main issue is how to trim the cache layer if it shouldn't be allowed to just grow indefinitely. Maybe that could even be a separate utility to just find oldest files and remove them at some watermark capacity level.
